File: go-leetcode/stack-and-queue/496.下一个更大元素 I.go
package leetcode

func nextGreaterElement(nums1 []int, nums2 []int) []int {
	ans := make([]int, len(nums1))
	for i := range ans {
		ans[i] = -1
	}
	stack := []int{}
	m := make(map[int]int)

	for i := 0; i < len(nums2); i++ {
		for len(stack) > 0 && nums2[i] > nums2[stack[len(stack)-1]] {
			top := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			m[nums2[top]] = nums2[i]
		}
		stack = append(stack, i)
	}

	for i, num := range nums1 {
		if val, ok := m[num]; ok {
			ans[i] = val
		}
	}

	return ans
}

File: python-leetcode/stack-and-queue/496.下一个更大元素 I.py
from typing import List


class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        res = [-1] * len(nums1)
        stack: List[int] = []
        mapping = {}

        for i in range(len(nums2)):
            while stack and nums2[i] > nums2[stack[-1]]:
                top_index = stack.pop()
                mapping[nums2[top_index]] = nums2[i]
            stack.append(i)

        for i in range(len(nums1)):
            if nums1[i] in mapping:
                res[i] = mapping[nums1[i]]

        return res
